From 83771e186d2abeccc38b0a5e6841b068499622ca Mon Sep 17 00:00:00 2001
From: Cheick Traore <cheick.traore@foss.st.com>
Date: Tue, 22 Oct 2024 11:36:55 +0200
Subject: [PATCH] serial: stm32: retrieve data stuck in DMA FIFO

When using DMA with usart peripheral to transfer an amount of data
which is not multiple of the DMA FIFO burst, last data may be lost
because they are still stuck in the DMA FIFO. Use "in_flight_bytes" from
dma_tx_state, which represents the amount of bytes stuck in DMA FIFO to
take these in flight bytes into account before pushing them to
framework.

Change-Id: I14489d1b39289da4409ca0137da46b9f5d0db5f7
Signed-off-by: Cheick Traore <cheick.traore@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/413162
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
Domain-Review: Amelie DELAUNAY <amelie.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 drivers/tty/serial/stm32-usart.c | 26 +++++++++++++++++++++-----
 1 file changed, 21 insertions(+), 5 deletions(-)

--- a/drivers/tty/serial/stm32-usart.c
+++ b/drivers/tty/serial/stm32-usart.c
@@ -113,6 +113,7 @@ static struct stm32_usart_info __maybe_u
 static void stm32_usart_stop_tx(struct uart_port *port);
 static void stm32_usart_transmit_chars(struct uart_port *port);
 static void __maybe_unused stm32_usart_console_putchar(struct uart_port *port, unsigned char ch);
+static int stm32_usart_rx_dma_start_or_resume(struct uart_port *port);
 
 static inline struct stm32_port *to_stm32_port(struct uart_port *port)
 {
@@ -648,18 +649,33 @@ static unsigned int stm32_usart_receive_
 {
 	struct stm32_port *stm32_port = to_stm32_port(port);
 	const struct stm32_usart_offsets *ofs = &stm32_port->info->ofs;
+	struct dma_tx_state *state;
 	enum dma_status rx_dma_status;
 	u32 sr;
 	unsigned int size = 0;
 
 	if (stm32_usart_rx_dma_started(stm32_port) || force_dma_flush) {
+		state = &stm32_port->rx_dma_state;
 		rx_dma_status = dmaengine_tx_status(stm32_port->rx_ch,
-						    stm32_port->rx_ch->cookie,
-						    &stm32_port->rx_dma_state);
+						    stm32_port->rx_ch->cookie, state);
 		if (rx_dma_status == DMA_IN_PROGRESS ||
 		    rx_dma_status == DMA_PAUSED) {
-			/* Empty DMA buffer */
-			size = stm32_usart_receive_chars_dma(port);
+			if (force_dma_flush && state->in_flight_bytes) {
+				/* Disable RX DMA to force in flight data is drained */
+				stm32_usart_rx_dma_terminate(stm32_port);
+				state->residue -= state->in_flight_bytes;
+				/* Empty DMA buffer */
+				size = stm32_usart_receive_chars_dma(port);
+				if (rx_dma_status == DMA_IN_PROGRESS) {
+					/* If restarting DMA fails, fall back to interrupt mode */
+					if (stm32_usart_rx_dma_start_or_resume(port))
+						size = stm32_usart_receive_chars_pio(port);
+				}
+			} else {
+				/* Empty DMA buffer */
+				size = stm32_usart_receive_chars_dma(port);
+			}
+
 			sr = readl_relaxed(port->membase + ofs->isr);
 			if (sr & USART_SR_ERR_MASK) {
 				/* Disable DMA request line */
@@ -1117,7 +1133,7 @@ static irqreturn_t stm32_usart_interrupt
 	/* Receiver timeout irq for DMA RX */
 	if (stm32_usart_rx_dma_started(stm32_port) && !stm32_port->throttled) {
 		spin_lock(&port->lock);
-		size = stm32_usart_receive_chars(port, false);
+		size = stm32_usart_receive_chars(port, (sr & USART_SR_RTOF));
 		uart_unlock_and_check_sysrq(port);
 		if (size)
 			tty_flip_buffer_push(tport);
